Store: Add Low Inventory widget to the dashboard

ryelle opened the pull request:
See #16732 – this PR adds a new widget to the dashboard, on both the no-orders & has-orders view, to show products with low or no inventory.

Low stock, on "has orders" dashboard, sits in next to the reports widget. This location might need rethinking if the stats widget is meant to replace the reports widget (#23344)
<img width="910" alt="screen shot 2018-03-15 at 4 52 55 pm" src="https://user-images.githubusercontent.com/541093/37490916-8017a750-2872-11e8-8a18-459cbc2846e0.png">

No stock on a product turns the border red, and this is on the "no orders" view, where it replaces the stats widget (unless there are no products).
<img width="906" alt="screen shot 2018-03-15 at 5 05 55 pm" src="https://user-images.githubusercontent.com/541093/37491128-292e0dd4-2873-11e8-936d-aa21209fa4c4.png">

This is the view when there are no products low or out of stock, maybe we could have an illustration here?
<img width="451" alt="screen shot 2018-03-15 at 4 41 10 pm" src="https://user-images.githubusercontent.com/541093/37490918-803f4ef4-2872-11e8-9c00-6f46b6631ff5.png">

This doesn't add the initial view (which I think instead should be the "inventory management is disabled" view), or the configuration options. I'll follow up with another PR for that.

**To test**

- View your dashboard, `/store/:site`
- If you have products, you should see this new widget
- Low or No stock is determined by the settings in `/wp-admin/admin.php?page=wc-settings&tab=products&section=inventory`
- You can change product stock quantity to see the different states: low stock, no stock, everything's OK

a8ci18n commented on client/extensions/woocommerce/app/dashboard/dashboard-widgets/inventory.js:
```diff
@@ -0,0 +1,176 @@
+					{ translate( 'Product' ) }
+				</TableItem>
+				<TableItem isHeader isTitle>
+					{ translate( 'Stock' ) }
```
Hi! I've found a possible matching string that has already been translated **17** times:
`translate( 'Stocks' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

ryelle commented:
I moved the widget to full width and added the image, but it looks even more sparse at full-width:
<img width="911" alt="screen shot 2018-03-16 at 11 14 55 am" src="https://user-images.githubusercontent.com/541093/37528617-77149ebc-290b-11e8-8d23-b7341ee9bfbd.png">

A product list looks OK at full width with the image:
<img width="907" alt="screen shot 2018-03-16 at 11 19 34 am" src="https://user-images.githubusercontent.com/541093/37528827-f693d6e4-290b-11e8-8ac4-10923e415bdf.png">

Another idea (that would clearly need more iteration) would be to drop it in the top row, and reconfigure the orders-to-process widget. When the reviews are showing up, it could be third/third/third, the inventory widget does OK at 1/3 width. When no reviews, it could be half/half, or two-thirds/third. With the stats widget, this revenue count might not be that useful, too (TBH I only just realized it's the revenue of pending orders, not like... revenue in the last [time period])

<img width="905" alt="screen shot 2018-03-16 at 11 25 27 am" src="https://user-images.githubusercontent.com/541093/37529178-d4a7ded0-290c-11e8-9fc0-d37928f60c24.png">

timmyc commented on client/extensions/woocommerce/app/dashboard/widgets/inventory-widget/index.js:
```diff
@@ -0,0 +1,185 @@
+		if ( site && site.ID ) {
+			this.props.fetchSettingsProducts( site.ID );
+			// Get all products
+			this.props.fetchProducts( site.ID, { offset: 0 } );
```
So it appears that `fetchProducts` utilizes the [DEFAULT_QUERY](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/woocommerce/state/sites/products/utils.js#L9) which limits this request to 10 products per page. Wondering if we should consider bumping up that per page style for this widget to try and cast a wider net and get more products that might be low inventory / out of stock.

It looks like the [products endpoint](http://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products) supports an `in_stock` query arg too - so i suppose another route we could do here is if `props. isLowStockEnabled` is falsey - we just change this query to only grab out of stock products for display.

ryelle replied:
The `offset: 0` triggers [a loop of product-fetches](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/woocommerce/state/data-layer/products/index.js#L138-L149), so the entire list of products is retrieved - but increasing that default per_page would mean fewer api requests so I'll do that.

a8ci18n commented on client/extensions/woocommerce/app/dashboard/widgets/inventory-widget/index.js:
```diff
@@ -0,0 +1,185 @@
+					{ translate( 'Product' ) }
+				</TableItem>
+				<TableItem isHeader isTitle>
+					{ translate( 'Stock' ) }
```
Hi! I've found a possible matching string that has already been translated **17** times:
`translate( 'Stocks' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

ryelle commented:
> The latest push is resulting in an error for me locally

Oops, I renamed the folder but forgot to update it in the dashboard. Fixed now!

ryelle commented:
I think I misunderstood the use of `isLowStockEnabled`, I think it's meant to be a toggle for whether the store owner is emailed when stock is low (at least, that's how it's used in [the mockup](https://camo.githubusercontent.com/09d019a3943eee5e3ff55367240a1b7794095210/68747470733a2f2f636c6475702e636f6d2f434d392d5067393976792e706e67)). I've removed the logic checking for that, since I think this widget should be displayed whether or not the store owner wants emails.

timmyc commented:
I think you have a great point there @ryelle - the two notification preferences around low stock / out of stock are currently only used to determine wether or not emails should be sent [#](https://github.com/woocommerce/woocommerce/blob/f5492cc50a1c4b9ba96edecb6ab55da93ee81c72/includes/class-wc-emails.php#L504-L547) during these events.

So the current implementation of showing this widget if `woocommerce_manage_stock === 'yes'` seems like the best way to handle it.

Grabbed the fixed branch and found a few things. FWIW I turned off the force scroll bars in osx :)

- When clicking on any product in the inventory alert widget, the link goes to http://calypso.localhost:3000/store/product/SITE/undefined
- The image seems really large to me in most instances, but especially on mobile - and at other times there seems to be a large amount of whitespace:

![lots-of-whitespace](https://user-images.githubusercontent.com/22080/37543135-e0e17c40-291d-11e8-945a-5ee46e4f0782.png)
![large-mobile](https://user-images.githubusercontent.com/22080/37543198-128c91ee-291e-11e8-9e88-e61fb5e9c858.png)
![warnings](https://user-images.githubusercontent.com/22080/37543233-2adbd87c-291e-11e8-8e8f-0cde9f09755d.png)

- on mobile, the list of products can mistakenly be scrolled itself - which can cut off some of the data and is just kind of funky:

![alerts](https://user-images.githubusercontent.com/22080/37543371-8fa4b4ea-291e-11e8-8c83-cd25d5a08593.gif)

jameskoster commented:
> Another idea (that would clearly need more iteration) would be to drop it in the top row, and reconfigure the orders-to-process widget.

That would be my preference. I think we should explore combining all of the action items in to a single widget. Obviously we'll have to do that separately like you said.

For now I think this is ok, but the image was only intended to be displayed when there are no stock notifications. On that note, if we're unhappy about how the widget looks when there are no notifications we could just hide it. The only reason to display it was so that the [settings](https://cloudup.com/cvDGHSKH9lB) were always accessible, but I don't think we're adding that here?

ryelle commented:
I rebased again to fix the test; also added commits to fix the `undefined` in the product link, and remove the image & fixed height from mobile screens, to prevent that double-scroll. It looks like this now:

| small screen, no alerts | small screen, has alerts |
| --- | --- |
| <img width="612" alt="screen shot 2018-03-19 at 11 01 31 am" src="https://user-images.githubusercontent.com/541093/37603411-e8da7390-2b64-11e8-9095-7f55fe6bf118.png"> | <img width="626" alt="has-alerts-small" src="https://user-images.githubusercontent.com/541093/37603034-09f8b7c2-2b64-11e8-87e1-d7cb68377172.png"> |

| large screen, no alerts | large screen, has alerts |
| --- | --- |
| <img width="907" alt="empty-large" src="https://user-images.githubusercontent.com/541093/37603202-76526e90-2b64-11e8-8325-8b541692c1df.png"> | <img width="907" alt="has-alerts-large" src="https://user-images.githubusercontent.com/541093/37603199-763795e8-2b64-11e8-93bf-d188931d3317.png"> |

> The only reason to display it was so that the settings were always accessible, but I don't think we're adding that here?

I'll have another PR up shortly with the settings, I just wanted this to get some review first. Now that I've addressed @timmyc's mobile feedback, are you OK with me merging this, design-wise?
